src/modules/sensors/vehicle_gps_position/VehicleGPSPosition.cpp

```diff
@@ -0,0 +1,742 @@
+	float alpha[GPS_MAX_RECEIVERS] = {};
+	float omega_lpf = 1.0f / fmaxf(_param_sens_gps_tau.get(), 1.0f);
+
+	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+		if (_gps_state[i].timestamp - _time_prev_us[i] > 0) {
+			// calculate the filter coefficient that achieves the time constant specified by the user adjustable parameter
+			float min_alpha = constrain(omega_lpf * 1e-6f * (float)(_gps_state[i].timestamp - _time_prev_us[i]),
+						    0.0f, 1.0f);
+
+			// scale the filter coefficient so that time constant is inversely proprtional to weighting
+			if (_blend_weights[i] > min_alpha) {
+				alpha[i] = min_alpha / _blend_weights[i];
+
+			} else {
+				alpha[i] = 1.0f;
+			}
+
+			_time_prev_us[i] = _gps_state[i].timestamp;
+		}
+	}
+
+	// Calculate a filtered position delta for each GPS relative to the blended solution state
+	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+		Vector2f offset;
+		get_vector_to_next_waypoint((_gps_state[i].lat / 1.0e7), (_gps_state[i].lon / 1.0e7),
+					    (_gps_blended_state.lat / 1.0e7), (_gps_blended_state.lon / 1.0e7), &offset(0), &offset(1));
+		_NE_pos_offset_m[i] = offset * alpha[i] + _NE_pos_offset_m[i] * (1.0f - alpha[i]);
+		_hgt_offset_mm[i] = (float)(_gps_blended_state.alt - _gps_state[i].alt) *  alpha[i] +
+				    _hgt_offset_mm[i] * (1.0f - alpha[i]);
+	}
+
+	// calculate offset limits from the largest difference between receivers
+	Vector2f max_ne_offset{};
+	float max_alt_offset = 0;
+
+	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+		for (uint8_t j = i; j < GPS_MAX_RECEIVERS; j++) {
+			if (i != j) {
+				Vector2f offset;
+				get_vector_to_next_waypoint((_gps_state[i].lat / 1.0e7), (_gps_state[i].lon / 1.0e7),
+							    (_gps_state[j].lat / 1.0e7), (_gps_state[j].lon / 1.0e7), &offset(0), &offset(1));
+				max_ne_offset(0) = fmaxf(max_ne_offset(0), fabsf(offset(0)));
+				max_ne_offset(1) = fmaxf(max_ne_offset(1), fabsf(offset(1)));
+				max_alt_offset = fmaxf(max_alt_offset, fabsf((float)(_gps_state[i].alt - _gps_state[j].alt)));
+			}
+		}
+	}
+
+	// apply offset limits
+	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+		_NE_pos_offset_m[i](0) = constrain(_NE_pos_offset_m[i](0), -max_ne_offset(0), max_ne_offset(0));
+		_NE_pos_offset_m[i](1) = constrain(_NE_pos_offset_m[i](1), -max_ne_offset(1), max_ne_offset(1));
+		_hgt_offset_mm[i] = constrain(_hgt_offset_mm[i], -max_alt_offset, max_alt_offset);
+	}
+}
+
+void VehicleGPSPosition::apply_gps_offsets()
+{
+	// calculate offset corrected output for each physical GPS.
+	for (uint8_t i = 0; i < GPS_MAX_RECEIVERS; i++) {
+		// Add the sum of weighted offsets to the reference position to obtain the blended position
+		double lat_deg_now = (double)_gps_state[i].lat * 1.0e-7;
+		double lon_deg_now = (double)_gps_state[i].lon * 1.0e-7;
+		double lat_deg_res, lon_deg_res;
+		add_vector_to_global_position(lat_deg_now, lon_deg_now, _NE_pos_offset_m[i](0), _NE_pos_offset_m[i](1), &lat_deg_res,
```

Comment:
I don't think we should modify the GPS individual readings; IMO, the blending logic should create a blended output from all the receivers but not adjust each GPS solution, why do we do that?

Reply:
@dakejahl @JacobCrabill any comment here? I don't believe I modified any of the logic when I did the original extraction in https://github.com/PX4/Firmware/pull/13584, although obviously it's been a while...

Reply:
Ok, looks like this isn't modified from what ekf2 is currently doing.

> I don't think we should modify the GPS individual readings; IMO, the blending logic should create a blended output from all the receivers but not adjust each GPS solution

I agree, storing some of this incremental state was one of my criticisms of the original gps blending work in ekf2, but it was a bit more confusing going between uORB `vehicle_gps_position` -> ecl/EKF `gps_message` and then back to `ekf_gps_position` (nearly identical to `vehicle_gps_position`). Let me see if I can eliminate it this time.